cmd/nvidia-dra-plugin/main.go

```diff
@@ -183,19 +183,24 @@ func StartPlugin(ctx context.Context, config *Config) error {
 		return fmt.Errorf("path for cdi file generation is not a directory: '%v'", config.flags.cdiRoot)
 	}
 
-	driver, err := NewDriver(ctx, config)
-	if err != nil {
-		return err
-	}
-
 	sigc := make(chan os.Signal, 1)
 	signal.Notify(sigc, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
-	<-sigc
 
-	err = driver.Shutdown(ctx)
+	var driver *driver
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+		if err := driver.Shutdown(); err != nil {
+			klog.Errorf("Unable to cleanly shutdown driver: %v", err)
+		}
+	}()
```

Comment:
Is there any reason that we we can't add this `defer` AFTER we have successfully created the driver? Except for the return on error there, there doesn't seem to be another point where this could be triggered otherwise.

Reply:
We can't unless we break it into 2 defers. The linter will complain if we don't call to cancel on all exit paths after the context it controls has been created. This includes the error path from the creation of the driver since it takes the context as a parameter.

Reply:
Actually I don't think breaking it into 2 defers would even work since they run last-in-first-out.

Reply:
OK. That's a sufficient explaination. We have to define it where it is due to the cancel context and not due to the driver specifically.